### PR TITLE
added plot maker for anaTuple

### PR DESCRIPTION
**Pull request by valeriadamante**

### CODE REVIEW

File: Common/skimAnatuple.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    for tau_idx in [1,2]:
+        df = df.Filter(f'tau{tau_idx}_pt>40 && abs(tau{tau_idx}_eta)<2.1')
+    df = df.Filter('DeltaR(tau1_eta,tau1_phi, tau2_eta,tau2_phi) > 0.5')
+    df = df.Filter('tau1_charge+tau2_charge==0')
+    return df
+
+def findInvMass(df):
+    for idx in [0,1]:
+        df = df.Define(f"tau{idx+1}_p4", f"ROOT::Math::LorentzVector<ROOT::Math::PtEtaPhiM4D<double>>(tau{idx+1}_pt,tau{idx+1}_eta,tau{idx+1}_phi,tau{idx+1}_mass)")
+    df = df.Define("tau_inv_mass", "(tau1_p4+tau2_p4).M()")
```
Comment (kandrosov) [inline Common/skimAnatuple.py:17]:
```suggestion
    df = df.Define(tautau_m_vis", "(tau1_p4+tau2_p4).M()")
```

File: AnaProd/MergeNtuples.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+                    lumiEventMap[lumi] = s ;
+                    eventMap[run]= lumiEventMap;
+                }*/
+            }
+            eventMap_mutex.lock();
+            static std::set<unsigned long long> s;
+            s.insert(event);
+            static lumiEventMapType lumiEventMap;
+            lumiEventMap[lumi] = s ;
+            eventMap[run]= lumiEventMap;
```
Comment (kandrosov) [inline AnaProd/MergeNtuples.py:34]: fix

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: Analysis/Producer.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import os
+import sys
+import ROOT
+
+if __name__ == "__main__":
+    sys.path.append(os.environ['ANALYSIS_PATH'])
+
+import Common.Utilities as Utilities
+import PlotKit.Plotter as Plotter
+import Common.skimAnatuple as skimAnatuple
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply Analysis/Producer.py:12]: ok

File: Analysis/Producer.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    "TT":["TT"],
+    "VBFToBulkGraviton":["VBFToBulkGraviton"],
+    "VBFToRadion":["VBFToRadion"],
+    "W":["W"],
+    #"data":["data"],
+}
+
+def MakeStackedPlot(df_dict, var):
+    hists = {}
+    plotter = Plotter.Plotter(page_cfg=page_cfg, page_cfg_custom=page_cfg_custom, hist_cfg=hist_cfg, inputs_cfg=inputs_cfg)
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply Analysis/Producer.py:33]: ok

File: Common/AnalysisTools.h [hunk 5]
```diff
@@ -106,6 +106,10 @@ T DeltaR(T eta1, T phi1, T eta2, T phi2){
   return std::hypot(dphi, deta);
 }
 
+template <typename T> int sgn(T val) {
+    return (T(0) < val) - (val < T(0));
+}
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply Common/AnalysisTools.h:111]: ok

File: AnaProd/MergeNtuples.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
+import ROOT
+ROOT.gInterpreter.Declare(
+    """
+    using lumiEventMapType = std::map < unsigned int, std::set <unsigned long long > >;
+    using eventMapType =  std::map < unsigned int ,  lumiEventMapType >;
+    using RunLumiEventSet = std::set < std::tuple < unsigned int, unsigned int, unsigned long long > > ;
+
```
Comment (kandrosov) [inline AnaProd/MergeNtuples.py:7]: still not:
```
print(ROOT.saveEvent(1, 2, 3))
print(ROOT.saveEvent(2, 1, 3))
print(ROOT.saveEvent(2, 2, 3))
```

```
True
True
False
```

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: Common/skimAnatuple.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import ROOT
+import os
+headers_dir = os.path.dirname(os.path.abspath(__file__))
+header_path_AnalysisTools = os.path.join(headers_dir, "AnalysisTools.h")
+ROOT.gInterpreter.Declare(f'#include "{header_path_AnalysisTools}"')
+
+def skimAnatuple(df):
+    for tau_idx in [1,2]:
+        df = df.Filter(f'tau{tau_idx}_pt>40 && abs(tau{tau_idx}_eta)<2.1')
+    df = df.Filter('DeltaR(tau1_eta,tau1_phi, tau2_eta,tau2_phi) > 0.5')
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply Common/skimAnatuple.py:10]: ok

File: Common/skimAnatuple.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    df = df.Filter('tau1_charge+tau2_charge==0')
+    return df
+
+def findInvMass(df):
+    for idx in [0,1]:
+        df = df.Define(f"tau{idx+1}_p4", f"ROOT::Math::LorentzVector<ROOT::Math::PtEtaPhiM4D<double>>(tau{idx+1}_pt,tau{idx+1}_eta,tau{idx+1}_phi,tau{idx+1}_mass)")
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply Common/skimAnatuple.py:16]: ok

File: Common/skimAnatuple.py (hunk 1 above)
Comment (valeriadamante) [reply Common/skimAnatuple.py:17]: ok

File: AnaProd/MergeNtuples.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+                s.insert(event);
+                eventMap[run][lumi]=s;
+            }
+        }
+        else{
+            std::set<unsigned long long> s;
+            s.insert(event);
+            eventMap[run][lumi]=s;
+        }
+            return true;
```
Comment (kandrosov) [inline AnaProd/MergeNtuples.py:34]: ok, now it looks to be working. Just cosmetic changes below
```suggestion
    auto& events = eventMap[run][lumi];
    if(events.find(event) != events.end())
        return false;
    events.insert(event);
    return true;
```

File: Analysis/Producer.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    "DY":["DY"],
+    "Other":["EWK", "ST", "TTT", "TTTT", "TTVH", "TTV", "TTVV", "TTTV", "VH", "VV", "VVV", "H", "HHnonRes", "TTHH", "ttH"],
+    #"GluGluToBulkGraviton":["GluGluToBulkGraviton"],
+    "QCD":["QCD"],
+    "TT":["TT"],
+    "GluGluToRadion":["GluGluToRadion"],
+    #"VBFToBulkGraviton":["VBFToBulkGraviton"],
+    # "VBFToRadion":["VBFToRadion"],
+    "W":["W"],
+    "data":["data"],
```
Comment (kandrosov) [inline Analysis/Producer.py:21]: use `data_unique` - which should be `data.root` on which applied `merge_tuple` filter

File: Analysis/Producer.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+class AnaSkimmer:
+    def defineP4(self):
+        for idx in [0,1]:
+            self.df = self.df.Define(f"tau{idx+1}_p4", f"ROOT::Math::LorentzVector<ROOT::Math::PtEtaPhiM4D<double>>(tau{idx+1}_pt,tau{idx+1}_eta,tau{idx+1}_phi,tau{idx+1}_mass)")
+            self.df = self.df.Define(f"b{idx+1}_p4", f"ROOT::Math::LorentzVector<ROOT::Math::PtEtaPhiM4D<double>>(b{idx+1}_pt,b{idx+1}_eta,b{idx+1}_phi,b{idx+1}_mass)")
+            self.df = self.df.Define(f"tau{idx+1}_seedingJet_p4", f"ROOT::Math::LorentzVector<ROOT::Math::PtEtaPhiM4D<double>>(tau{idx+1}_seedingJet_pt,tau{idx+1}_seedingJet_eta,tau{idx+1}_seedingJet_phi,tau{idx+1}_mass)")
+
+    def createInvMass(self):
+        self.df = self.df.Define("tautau_m_vis", "(tau1_p4+tau2_p4).M()")
+        self.df = self.df.Define("tautau_seedingJet_m_vis", "(tau1_seedingJet_p4+tau2_seedingJet_p4).M()")
```
Comment (kandrosov) [inline Analysis/Producer.py:35]:
```suggestion
```

File: Analysis/Producer.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+}
+signals = ["GluGluToBulkGraviton", "GluGluToRadion", "VBFToBulkGraviton", "VBFToRadion"]
+deepTauYears = {'v2p1':'2017','v2p5':'2018'}
+
+class AnaSkimmer:
+    def defineP4(self):
+        for idx in [0,1]:
+            self.df = self.df.Define(f"tau{idx+1}_p4", f"ROOT::Math::LorentzVector<ROOT::Math::PtEtaPhiM4D<double>>(tau{idx+1}_pt,tau{idx+1}_eta,tau{idx+1}_phi,tau{idx+1}_mass)")
+            self.df = self.df.Define(f"b{idx+1}_p4", f"ROOT::Math::LorentzVector<ROOT::Math::PtEtaPhiM4D<double>>(b{idx+1}_pt,b{idx+1}_eta,b{idx+1}_phi,b{idx+1}_mass)")
+            self.df = self.df.Define(f"tau{idx+1}_seedingJet_p4", f"ROOT::Math::LorentzVector<ROOT::Math::PtEtaPhiM4D<double>>(tau{idx+1}_seedingJet_pt,tau{idx+1}_seedingJet_eta,tau{idx+1}_seedingJet_phi,tau{idx+1}_mass)")
```
Comment (kandrosov) [inline Analysis/Producer.py:31]:
```suggestion
    def defineP4(self, name):
        self.df = self.df.Define(f"{name}_p4", f"ROOT::Math::LorentzVector<ROOT::Math::PtEtaPhiM4D<double>>({name}_pt,{name}_eta,{name}_phi,{name}_mass)")
        
    def defineAllP4(self):
        for idx in [0,1]:
            self.defineP4(f"tau{idx+1}")
            self.defineP4(f"b{idx+1}")
            self.defineP4(f"tau{idx+1}_seedingJet")
```

File: Analysis/Producer.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+        self.df = df
+        self.deepTauVersion = 'v2p1'
+        self.deepTauYear = deepTauYears[self.deepTauVersion]
+        self.defineP4()
+        self.createInvMass()
+
+
+    def skimAnatuple(self):
+        for tau_idx in [1,2]:
+            self.df = self.df.Filter(f'tau{tau_idx}_pt>40 && abs(tau{tau_idx}_eta)<2.1')
```
Comment (kandrosov) [inline Analysis/Producer.py:48]:
```suggestion
```

File: Analysis/Producer.py [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+
+
+
+
+def createHistograms(df_dict, var):
+    hists = {}
+    x_bins = Utilities.ListToVector(plotter.hist_cfg[var]['x_bins'], "double")
+    model = ROOT.RDF.TH1DModel("", "", x_bins.size()-1, x_bins.data())
+    for sample in df_dict.keys():
+        weight_names=[]
```
Comment (kandrosov) [inline Analysis/Producer.py:61]: if sample is data define weight 1.

File: Analysis/Producer.py [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+
+if __name__ == "__main__":
+    sys.path.append(os.environ['ANALYSIS_PATH'])
+
+weights_to_apply = ["weight_Central","weight_tau1_TrgSF_ditau_Central","weight_tau2_TrgSF_ditau_Central", "weight_tauID_Central",]
+files= {
+    "DY":["DY"],
+    "Other":["EWK", "ST", "TTT", "TTTT", "TTVH", "TTV", "TTVV", "TTTV", "VH", "VV", "VVV", "H", "HHnonRes", "TTHH", "ttH"],
+    #"GluGluToBulkGraviton":["GluGluToBulkGraviton"],
+    "QCD":["QCD"],
```
Comment (kandrosov) [inline Analysis/Producer.py:15]: - don't include MC QCD and TTHH
- add HHnonRes as a separate entry

File: Analysis/Producer.py [hunk 16]
```diff
@@ -0,0 +1,120 @@
+import ROOT
+import Common.Utilities as Utilities
+import sys
+import os
+
+
+if __name__ == "__main__":
+    sys.path.append(os.environ['ANALYSIS_PATH'])
+
+weights_to_apply = ["weight_Central","weight_tau1_TrgSF_ditau_Central","weight_tau2_TrgSF_ditau_Central", "weight_tauID_Central",]
```
Comment (kandrosov) [inline Analysis/Producer.py:10]: for v2p5 use Paola's tau ID

File: Analysis/Producer.py [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+        self.deepTauYear = deepTauYears[self.deepTauVersion]
+        self.defineP4()
+        self.createInvMass()
+
+
+    def skimAnatuple(self):
+        for tau_idx in [1,2]:
+            self.df = self.df.Filter(f'tau{tau_idx}_pt>40 && abs(tau{tau_idx}_eta)<2.1')
+            self.df = self.df.Filter(f'tau{tau_idx}_idDeepTau{self.deepTauYear}{self.deepTauVersion}VSjet >= {Utilities.WorkingPointsTauVSjet.Medium.value}')
+        self.df = self.df.Filter('std::copysign(tau1_charge, tau2_charge)<0')
```
Comment (kandrosov) [inline Analysis/Producer.py:50]:
```suggestion
        self.df = self.df.Filter('tau1_charge * tau2_charge <0')
```

File: Analysis/Producer.py [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+        if(sample in signals):
+            for input in inputs_cfg_dict:
+                name = input['name']
+                if(name == sample):
+                    input['title']+= f"mass {args.mass}"
+            anaskimmer.df = anaskimmer.df.Filter(f"X_mass=={args.mass}")
+        anaskimmer.skimAnatuple()
+        dataframes[sample] = anaskimmer.df
+
+    for var in args.vars.split(','):
```
Comment (kandrosov) [inline Analysis/Producer.py:114]: create histograms for all vars without calling `GetValue`. Then do another loop to plot.

File: Analysis/Producer.py [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+        hists = createHistograms(dataframes, var)
+        for sample in hists.keys():
+            hist = hists[sample]
+            hists[sample] = hist.GetValue()
+        plotter.plot(var, hists, f"output/plots/{var}_XMass{args.mass}.pdf")
```
Comment (kandrosov) [inline Analysis/Producer.py:119]: category label is missing

File: Common/skimAnatuple.py [hunk 20]
```diff
@@ -0,0 +1,18 @@
+import ROOT
```
Comment (kandrosov) [inline Common/skimAnatuple.py:1]: remove this file

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: Analysis/Producer.py (hunk 11 above)
Comment (valeriadamante) [reply Analysis/Producer.py:35]: ok

File: Analysis/Producer.py (hunk 12 above)
Comment (valeriadamante) [reply Analysis/Producer.py:31]: ok

File: Analysis/Producer.py (hunk 13 above)
Comment (valeriadamante) [reply Analysis/Producer.py:48]: ok

File: Analysis/Producer.py (hunk 17 above)
Comment (valeriadamante) [reply Analysis/Producer.py:50]: ok

File: Common/skimAnatuple.py (hunk 20 above)
Comment (valeriadamante) [reply Common/skimAnatuple.py:1]: ok

File: config/plot/histograms.yaml [hunk 21]
```diff
@@ -0,0 +1,15 @@
+tau1_pt:
+  x_bins: [0, 40, 60, 80, 100, 150, 200, 250, 300]
+  x_title: tau_1 p_{T} (GeV)
+  y_title: Events
+  use_log_y: False
+  max_y_sf: 1.5
+  divide_by_bin_width: False
+
+tautau_m_vis:
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply config/plot/histograms.yaml:9]: ok

File: Common/skimAnatuple.py (hunk 20 above)
Comment (valeriadamante) [reply Common/skimAnatuple.py:1]: ok

File: Analysis/Producer.py (hunk 19 above)
Comment (valeriadamante) [reply Analysis/Producer.py:119]: ok

File: Analysis/Producer.py (hunk 14 above)
Comment (valeriadamante) [reply Analysis/Producer.py:61]: ok

File: Analysis/Producer.py (hunk 15 above)
Comment (valeriadamante) [reply Analysis/Producer.py:15]: ok

File: Analysis/Producer.py [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+    import argparse
+    import PlotKit.Plotter as Plotter
+    import yaml
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--period', required=False, type=str, default = 'Run2_2018')
+    parser.add_argument('--version', required=False, type=str, default = 'v2_deepTau_v2p1')
+    parser.add_argument('--vars', required=False, type=str, default = 'tau1_pt')
+    parser.add_argument('--mass', required=False, type=int, default=500)
+    args = parser.parse_args()
+    sys.path.append(os.environ['ANALYSIS_PATH'])
```
Comment (kandrosov) [inline Analysis/Producer.py:96]:
```suggestion
```
(already done above)

File: Analysis/Producer.py [hunk 23]
```diff
@@ -0,0 +1,142 @@
...
+        anaskimmer.skimAnatuple()
+        dataframes[sample] = anaskimmer.df
+
+    defineWeights(dataframes)
+    all_histograms = {}
+    vars = args.vars.split(',')
+    for var in vars:
+        hists = createHistograms(dataframes, var)
+        for sample in hists.keys():
+            hist = hists[sample]
```
Comment (kandrosov) [inline Analysis/Producer.py:132]:
```suggestion
```

File: Analysis/Producer.py [hunk 24]
```diff
@@ -0,0 +1,142 @@
+import ROOT
+import Common.Utilities as Utilities
+import sys
+import os
+
+
+if __name__ == "__main__":
+    sys.path.append(os.environ['ANALYSIS_PATH'])
```
Comment (kandrosov) [inline Analysis/Producer.py:8]: are you sure it is needed? `import Common.Utilities as Utilities` works on L2...

File: config/plot/histograms.yaml [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+  x_bins: [0, 40, 80,120, 160,200, 250, 300, 350, 400, 450, 500, 1000]
+  x_title: \tau\tau m_{inv} (GeV)
```
Comment (kandrosov) [inline config/plot/histograms.yaml:11]:
```suggestion
  x_title: M^{\tau\tau}_{vis} (GeV)
```

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: Analysis/Producer.py (hunk 16 above)
Comment (kandrosov) [reply Analysis/Producer.py:10]: moved to the next PR

File: Analysis/Producer.py (hunk 24 above)
Comment (valeriadamante) [reply Analysis/Producer.py:8]: going to check

File: config/plot/histograms.yaml (hunk 25 above)
Comment (valeriadamante) [reply config/plot/histograms.yaml:11]: ok

**Review by kandrosov — APPROVED** [review]

**Pull request merged**